Implement drill up/down the hierarchy field

mereghost opened the pull request:
#### ⚠️ Related WP: [OP#57819](https://community.openproject.org/work_packages/57819)

## What?

We want to be able to navigate to (and back to) the different branches of the hierarchy custom field and, of course, do the normal CRUD operations over the branch child nodes.

## How?

This PR makes a lot of changes on the way we normally have been doing this, by going in a more `turbo_frames` first approach.

First it makes the turbo frame use the `morph` refresh mechanism and also the action `advance` so that the browser location and history are updated on every navigation. This required some tweaks as the `Primer::Breadcrumbs` don't allow us to pass extra attributes to its hyperlinks (besides target and href).

The result is a clean approach with a more standard RoR view approach and following the hotwire guidelines.

There's a couple of interesting pieces that I learned around this: for turbo frame responses a specific layout needs to be used. This ensures a minimalist response and focuses on just the frame being acted upon.

## Known issues

While you can have multiple items in edit mode and update them separately, the moment you click on new, the entire state resets. This is due to the way we render the form inside the border-box footer needing the re-render of the entire component.

There's no easy way around it due to the usage of the primer components. We would need to be able to somehow generate just a HTML snipped of a single row of a `BorderBox` so that we could append/prepend/update it.

This will probably become very relevant the moment we try to make the **Insert Above** or **Insert Below** actions.

### Extra Notes
Special thanks for @akabiru that put up with my silly and ranty questions and to @brunopagno that helped me clean this mess. <3

mereghost committed:
Make labels into links

Unlocks the necessary routes, with an extra gimmick

Creates a show view, to allow rendering of singluar branches

Wraps the main component in a turbo-frame so we can use the advance action

Kharonus commented on app/components/admin/custom_fields/hierarchy/item_component.rb:
```diff
@@ -45,6 +45,10 @@ def short_text
         end
 
+        def children_count
+          pluralize(@hierarchy_item.children.count, "sub-item")
```
🟡 of course in the end we need i18n here to pluralize this stuff. You know the patterns?

mereghost commented on app/views/admin/custom_fields/hierarchy/items/index.html.erb:
```diff
@@ -31,4 +31,6 @@ See COPYRIGHT and LICENSE files for more details.
 <%= render(Admin::CustomFields::EditFormHeaderComponent.new(custom_field: @custom_field, selected: :items)) %>
 
-<%= render(Admin::CustomFields::Hierarchy::ItemsComponent.new(custom_field: @custom_field)) %>
+<%= turbo_frame_tag :hierarchy_list, data: { turbo_action: :advance } do %>
```
`advance` needed to be added here due to the `Breadcrumbs` not allowing us to pass down tags to the `a` elements.

mereghost committed:
Updates components and controller to deal with turbo updates and navigation down the tree

Fix a bug when returning errors on the form

Allows editing deeply nested items.

Moves edit to work with only turbo frames

Remove unused parameters

Kharonus commented on config/locales/en.yml:
```diff
@@ -254,6 +254,11 @@ en:
         remember_items_and_projects: "Remember to set items and projects in the respective tabs for this custom field."
+      hierarchy:
+        subitems:
+          zero: no sub-items
```
🟡 "zero" is currently a dead locale key, as those are not rendered. I'd rather remove it and add it, once we need this text.

mereghost replied:
This is just protection against weird cases. :P

Kharonus commented on app/components/admin/custom_fields/hierarchy/items_component.rb:
```diff
@@ -35,20 +35,44 @@ class ItemsComponent < ApplicationComponent
-          # TODO: This must be context aware (breadcrumbs)
-          @custom_field.hierarchy_root.children
+        def item_header
+          render(Primer::Beta::Breadcrumbs.new) do |loaf|
```
❤️ perfect naming of variables

Kharonus commented on app/controllers/admin/custom_fields/hierarchy/items_controller.rb:
```diff
@@ -73,45 +85,43 @@ def destroy
 
-        def find_model_object(object_id = :custom_field_id)
-          super
+        def find_model_object
```
🟡 if we do not call super, lets maybe rename the method to `find_custom_field`.

mereghost replied:
This leads to other errors. This dude here does a bunch of stuff for the sidebars etc.

Kharonus commented on app/services/custom_fields/hierarchy/hierarchical_item_service.rb:
```diff
@@ -84,7 +84,7 @@ def delete_branch(item:)
       # @return [Success(Array<CustomField::Hierarchy::Item>)]
       def get_branch(item:)
-        Success(item.ancestors.reverse)
+        Success(item.self_and_ancestors.reverse)
```
🟢 nice

Kharonus reviewed (commented):
some first feedback

brunopagno commented on app/components/admin/custom_fields/hierarchy/items_component.html.erb:
```diff
@@ -28,42 +28,33 @@ See COPYRIGHT and LICENSE files for more details.
 
 <%=
-  component_wrapper(tag: "turbo-frame") do
+  component_wrapper(tag: 'turbo-frame', refresh: :morph, data: { turbo_action: :advance }) do
```
This is totally out of context, but it bothers me a bit that our linter does not lint the code in `.html.erb` files. Notice here the single quotes~

I feel the urge to fix this :sweat_smile:

mereghost replied:
There's [erb_lint](https://github.com/Shopify/erb_lint) that Shopify ressurected recently. Might be worth taking a look at.

brunopagno commented on app/components/admin/custom_fields/hierarchy/item_component.rb:
```diff
@@ -35,31 +35,33 @@ class ItemComponent < ApplicationComponent
-          super
-          @custom_field = custom_field
-          @hierarchy_item = hierarchy_item
+        def initialize(item:, edit_item_form_data: { show: false })
```
I am thinking we could also change the `edit_item_form_data` parameter to be just a boolean, since we don't pass down other attributes along anymore.

e.g.
```ruby
def initialize(item:, show_edit_form: false)
```

Possibly we can do that at a separate moment, though

Kharonus commented on app/components/admin/custom_fields/hierarchy/item_component.html.erb:
```diff
@@ -28,42 +28,44 @@ See COPYRIGHT and LICENSE files for more details.
           item_information.with_column(mr: 2) do
-            render(Primer::Beta::Text.new(font_weight: :bold)) do
-              @hierarchy_item.label
+            render(Primer::Beta::Link.new(href: custom_field_item_path(model.root.custom_field_id, model), underline: false)) do
```
🟡 minor: can we use `@root` instead of `model.root` if we expose it?

Kharonus commented on app/components/admin/custom_fields/hierarchy/items_component.rb:
```diff
@@ -35,19 +35,42 @@ class ItemsComponent < ApplicationComponent
-        def items
-          # TODO: This must be context aware (breadcrumbs)
-          @custom_field.hierarchy_root.children
+        def branch_root = model
```
🟡 this one is unused, isn't it? Can we remove it?

Kharonus commented on app/contracts/custom_fields/hierarchy/update_item_contract.rb:
```diff
@@ -45,12 +45,7 @@ class UpdateItemContract < Dry::Validation::Contract
-          .exists?
-
-        if label_already_exists
+        if values[:item].siblings.where(label: value).any?
```
🎉 nice

Kharonus reviewed (approved):
Nice one, changes are way fewer then I'd expected.

Only have two non-blocking issues.

Also I added [this bug](https://community.openproject.org/wp/58700) for the left margin. It seems a rather easy fix in CSS, but I won't inject it in your PR, but simply add a small bugfix after it.

mereghost committed:
Incorporates @Kharonus feedback
